src/basic/blockdev-util.c

```diff
@@ -144,15 +145,17 @@ int block_get_originating(dev_t dt, dev_t *ret) {
                          * different physical devices, and we don't support that. */
                         if (!streq(a, b))
                                 return -ENOTUNIQ;
+                } else {
```

Comment:
There's something I'm not quite following in this if/else block - before the change, `found` was updated on each iteration of the loop. But now, it's only updated on the first iteration.
So won't the logic be different now, apart from the possible invalid struct dirent issue? Or am I missing something?

Reply:
so, this goes through the slaves/ dir in sysfs of a dm-crypt or dm-verity device and tries to find the backing block device of it. There is usually one, but in case of dm-verity there might be two: the data device and the device with the verity hashes. In the latter case we are OK with that for thi function as long as the referneced devices themselves are partitions of the same whole block device. This is what the loops checks. Now previously we'd look at each entry and compare it with the previous we encountered, and fail once we see two such devices not backed by the same whole block device. With this patch this is slightly altered: we compare with the first whole block device we found. Why the change? only to simplify things: if we wanted to do things exactly as before we'd have to strdup() each path we look at at each iteration, while with the new code we only need to do this at the first one.

The net effect is really the same: if you check if all elements in a set are the same in some regard it doesn't matter if you compare each element `e1`, `e2`, `e3` … `en` with the first element `e0` and abort one you find them non-matching or if  you compare each element `e(i)` with `e(i-1)`. You'll come to the same results.

if you follow what i mean.

or in other words: if we find n devices, then calling strdup() n times is ugly, if 1 time suffices too. assigning a pointer n times or 1 times is much less of an issue though.

Reply:
Thanks, that makes sense now, it wasn't immediately obvious to me.

Maybe change the last line of the comment above the path_join to something like:

> Hence, let's verify this by iterating over every node in the 'slaves/' directory and comparing them with the first that gets returned by 'dirent', to ensure they all point to the same device.

Also maybe it would also be clearer (I know it was already like this and it's not part of this change) if the `found` variable was renamed to `first` ? It's not really something that was "searched" for, it's just whatever is returned first.

Reply:
all addressed. ptal